qubes-demo-agent/src/main.rs

```diff
@@ -0,0 +1,116 @@
+use qubes_gui_agent_proto::DaemonToAgentEvent as Event;
+use std::convert::TryInto;
+use std::os::unix::io::AsRawFd as _;
+use std::task::Poll;
+
+fn main() -> std::io::Result<()> {
+    let (width, height) = (0x200, 0x100);
+    let mut connection = qubes_gui_gntalloc::new(0).unwrap();
+    let (mut vchan, conf) = qubes_gui_client::Client::agent(0).unwrap();
+    // we now have a vchan 🙂
+    println!("🙂 Somebody connected to us, yay!");
+    println!("Configuration parameters: {:?}", conf);
+    println!("Creating window");
+    let window = 50.try_into().unwrap();
+    let create = qubes_gui::Create {
+        rectangle: qubes_gui::Rectangle {
+            top_left: qubes_gui::Coordinates { x: 50, y: 400 },
+            size: qubes_gui::WindowSize { width, height },
+        },
+        parent: None,
+        override_redirect: 0,
+    };
+    vchan.send(&create, window).unwrap();
+    let mut buf = connection.alloc_buffer(width, height).unwrap();
+    let mut shade = vec![0xFF00u32; (width * height / 2).try_into().unwrap()];
+    buf.write(
+        qubes_castable::as_bytes(&shade[..]),
+        (width * height).try_into().unwrap(),
+    );
+    vchan
+        .send_raw(buf.msg(), window, qubes_gui::MSG_WINDOW_DUMP)
+        .unwrap();
+    let title = b"Qubes Demo Rust GUI Agent";
+    let mut title_buf = [0u8; 128];
+    title_buf[..title.len()].copy_from_slice(title);
+    vchan
+        .send_raw(&mut title_buf, window, qubes_gui::MSG_SET_TITLE)
+        .unwrap();
+    vchan
+        .send(
+            &qubes_gui::MapInfo {
+                override_redirect: 0,
+                transient_for: 0,
+            },
+            window,
+        )
+        .unwrap();
+    vchan.wait();
+    loop {
+        let (window, e) = loop {
+            match vchan.read_header().map(Result::unwrap) {
+                Poll::Pending => vchan.wait(),
+                Poll::Ready((hdr, body)) => match Event::parse(hdr, body).unwrap() {
+                    None => {}
+                    Some(ev) => break ev,
+                },
+            }
+        };
```

Comment:
I'm trying to understand the underlying mechanics of the vchan API, so I can understand what options we have for moving low-level details like these to a lower level of abstraction. Why does this loop only need `vchan.wait()`, while the outer loop uses `libc::poll`?

Reply:
Does this mean that when we receive a partial message, this loop will busy-wait for the rest? It's also worrying when it comes to what exactly the `wait`s will acknowledge in that situation.

Reply:
"wait" acknowledges an event - if there is no pending one, it will wait for one (that's actually the origin of the function name - without poll, your main loop would just use "wait" for waiting for events). 
As for what it actually acknowledges - an event is "something has changed" - on the low level you need to check yourself whether new data is available (since previous "wait" call) or some more space in the ring buffer is available (if you care about this case). I'm not sure how this is built into abstraction you see here...

BTW, for the vchan protocol itself, you may want to read https://roscidus.com/blog/blog/2019/01/01/using-tla-plus-to-understand-xen-vchan/